src/gui/attributetable/qgsattributetablemodel.cpp

```diff
@@ -967,3 +1007,13 @@ const QgsFeatureRequest &QgsAttributeTableModel::request() const
 {
   return mFeatureRequest;
 }
+
+
+void QgsAttributeTableModel::sliderCheck( int position )
```

Comment:
Instead of depending on the movement of the slider, what we should instead be doing is using calls to QgsAttributeTableModel::data to detect which features are being requested by the view, and fetching the batches when a previously-unfetched feature is required. This removes the fragile approach used here which is mixing gui code with model code and which is the source of the issues you've identified.

I actually believe all this code is already in place, and it's how the attribute table used to work in very old QGIS versions. Features where fetched from the provider only when the model required them. I believe this was changed when sorting support was added to the table, as that required a complete load of the table in advance. I would suggest investigating whether disabling the initial full table load still falls back to the load-features-on-demand approach, and whether all that code is still in place and working correctly. Then we can investigate possible approaches to avoid the initial bulk-load when sorting is required.

Reply:
I have commented out the calls to loadLayer and the only thins that results are headers. I don't think the former code exists.

As far as relying on ::data to detect needed features, I am not sure how it would be possible. The table viewer by default doesn't request specific features. As this tool is more used to simply explorer the content of a layer and modify that content, there is no requests that are made usually. The only approach where I can see this being of use if when a user selects a layer and if we want that selected layer to appear in the table.

I'm also not sure if calls to data are being made, possibly these are only for display/rendering?

Reply:
@nyalldawson ok I did some test using QgsAttributeModel::data to see how the data is structured. Basically nothing is ordered sequentially. I have no clue why and this is especially the case with the dual view (I still need to patch it more).

Even in the base model view, number have a random number. Using my pointer and around row 20 to 31, The row value of the model index goes something like: 33,9,46,2,21,68..

Using this as a threshold would mean that the user can be randomly mousing over a cell and trigger the database to load another batch even if he is in the first 10 rows, or just go on a random hunt for the magic cells that will trigger another batch to load by mousing over everything. This is a far worse option than using the GUI.

Also I have checked the code from 2013 and the table that was loaded was based on a rectangle or selection but I couldn't identify the part that loaded the rest of the table when needed.

Reply:
@roya0045 you'll need to do some research in the qt model/view infrastructure. It's a mental twist at first, but a useful concept to understand fully. Unfortunately the qgis attribute table is NOT a good starting point for learning models... There's just too much happening here! But, short answer is, we **cannot** use the current approach of mixing the gui portions (the view) with the underlying model data. You've already bumped into some limitations of this approach, and we'll hit many more (probably unsolvable) ones in future if we go down this track (and the last thing we want is to make these classes more complex and confusing! 😱😜). So we need to find an approach which relies solely on changes in the model and/or attribute table cache only...

Reply:
I'll see what I can do come up with. Thought I don't quite like how often the data call is actually used in qt models ( as stated in my symbology PR), this make things more noisy. I also feel like for such an approach using ui as a trigger since it doesn't care about the content but only the position. Though I think that the issue I encountered are more related to the model. The error values are only element that are not longer present. The only wrong behaviour that I saw was due to the fact that the wrong function was called when too many features/row were deleted and that function didn't refresh the content of the table, and that call was inconsistent. I would not link this to interweaving the UI and model and the UI check seem to be resistant to the tests that I carried out.  Closing and reopening the table would close the only occasional issue I encountered. Further testing and on a bigger scale might help to show if this approach is truly bad.

As it is mostly aimed at interface element, I'm confused as to why the solution must not imply ui given that this part of the code does not appear to be problematic.

We can go over this later, I would appreciate if my more mature PR could be discussed beforehand. In the meantime I will see what I can do.

Reply:
Thanks @roya0045! I'd really encourage you to dig into how models are used elsewhere in qgis and use this to get a good grounding in the concept. They took me quite some time to get my head around at first! Good ones to study as examples would be QgsMapLayerModel and QgsFieldModel